journalbeat/reader/journal.go

```diff
@@ -179,23 +186,65 @@ func (r *Reader) seek(cursor string) {
 // Next waits until a new event shows up and returns it.
 // It blocks until an event is returned or an error occurs.
 func (r *Reader) Next() (*beat.Event, error) {
+	event, err := r.readEvent()
+	if err != nil {
+		return nil, err
+	}
+	if event != nil {
+		return event, nil
+	}
+
+	return r.waitUntilNewEventOrError()
+}
+
+func (r *Reader) waitUntilNewEventOrError() (*beat.Event, error) {
 	for {
 		select {
 		case <-r.done:
 			return nil, nil
-		default:
-			event, err := r.readEvent()
-			if err != nil {
-				return nil, err
-			}
+		case c := <-r.changesChan:
+			switch c {
+			// no changes
+			case sdjournal.SD_JOURNAL_NOP:
+			// new entries are added or the journal has changed (e.g. vacuum, rotate)
+			case sdjournal.SD_JOURNAL_APPEND, sdjournal.SD_JOURNAL_INVALIDATE:
+				event, err := r.readEvent()
+				if err != nil {
+					return nil, err
+				}
+
+				if event == nil {
+					r.backoff.Wait()
+					continue
+				}
+
+				r.backoff.Reset()
+				return event, nil
+			default:
+				if c < 0 {
+					return nil, fmt.Errorf("error while waiting for event: %+v", syscall.Errno(-c))
+				}
 
-			if event == nil {
-				r.backoff.Wait()
-				continue
+				r.logger.Errorf("Unknown return code from Wait: %d\n", c)
 			}
+		}
+	}
+}
+
+func (r *Reader) waitForChange() {
+	for {
+		// try to return before waiting for event
+		select {
+		case <-r.done:
+			return
+		default:
+		}
 
-			r.backoff.Reset()
-			return event, nil
+		c := r.journal.Wait(100 * time.Millisecond)
```

Comment:
Note: It's not too bad to wait for a few 100 millis, and I don't think we should change it here. But I'd prefer if we wouldn't have to wait. The Wait call uses sd_journal_wait, which itself uses ppoll. One should not close a file descriptor a poll operation is active on, but an alternate wait-without-wakeup-schema could be created by wasting a second polling file descriptor:
- create close-signaling file descriptor (e.g. create pipe, wastes 2 FDs)
- get journald fd via sd_journal_get_fd (it's the same journald polls on)
- poll on close-fd + journal-fd
- for handling async shutdown trigger close-fd (e.g. send some bytes)
- check who did send poll signal and handle accordingly